src/providers/vmware/mod.rs

```diff
@@ -0,0 +1,91 @@
+//! VMWare provider.
+//!
+//! This uses the guest->host backdoor protocol for introspection.
+
+use std::collections::HashMap;
+
+use error_chain::bail;
+use openssh_keys::PublicKey;
+use slog_scope::warn;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+/// Guestinfo key for network kargs.
+static KARGS_NM_INITRD: &str = "guestinfo.exp.org.freedesktop.NetworkManager.initrd";
+
+/// Path to fragment file with network kargs.
+static NET_KARGS_PATH: &str = "/etc/cmdline.d/60-afterburn-net-kargs.conf";
+
+/// VMWare provider.
+#[derive(Clone, Debug)]
+pub struct VmwareProvider {
+    /// Extra kargs for early network setup.
+    net_kargs: Option<String>,
+}
+
+impl VmwareProvider {
+    pub fn try_new() -> Result<Self> {
+        if !vmw_backdoor::is_vmware_cpu() {
+            bail!("not running on VMWare CPU");
+        }
+
+        let mut backdoor = vmw_backdoor::probe_backdoor()?;
+        let mut erpc = backdoor.open_enhanced_chan()?;
+        let net_kargs = Self::get_net_kargs(&mut erpc)?;
+
+        let provider = Self { net_kargs };
+        Ok(provider)
+    }
+
+    fn get_net_kargs(erpc: &mut vmw_backdoor::EnhancedChan) -> Result<Option<String>> {
+        let guestinfo = erpc
+            .get_guestinfo(KARGS_NM_INITRD.as_bytes())
+            .chain_err(|| "failed to retrieve network kargs")?
+            .map(|bytes| String::from_utf8_lossy(&bytes).into_owned());
+        Ok(guestinfo)
+    }
+}
+
+impl MetadataProvider for VmwareProvider {
+    fn rd_net_kargs(&self) -> Result<()> {
+        use std::io::Write;
+
+        let kargs = match &self.net_kargs {
+            Some(val) => val,
+            None => return Ok(()),
+        };
+        let mut fp = std::fs::File::create(NET_KARGS_PATH)
+            .chain_err(|| format!("failed to create {}", NET_KARGS_PATH))?;
+        fp.write_all(kargs.as_bytes())?;
+        fp.write(&[b'\n'])?;
+        Ok(())
+    }
+
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        Ok(HashMap::new())
+    }
+
+    fn hostname(&self) -> Result<Option<String>> {
```

Comment:
We can implement this right?  See https://github.com/openshift/machine-config-operator/pull/1579

Reply:
I've replied there. The TLDR is that we can but we shouldn't.
This is for platforms where either:
 * network auto-configuration is performed via DHCP and a cloud-channel for the hostname
 * network auto-configuration happens via some custom protocol, which includes the hostname

Other cases are owned already by 1) dracut kargs and 2) Ignition.